src/coreclr/vm/methodtable.cpp

```diff
@@ -4222,9 +4223,20 @@ OBJECTREF MethodTable::GetManagedClassObject()
 
         REFLECTCLASSBASEREF  refClass = NULL;
         GCPROTECT_BEGIN(refClass);
-        refClass = (REFLECTCLASSBASEREF) AllocateObject(g_pRuntimeTypeClass);
 
-        LoaderAllocator *pLoaderAllocator = GetLoaderAllocator();
+        LoaderAllocator* pLoaderAllocator = GetLoaderAllocator();
+        if (!pLoaderAllocator->CanUnload())
+        {
+            // Allocate RuntimeType on a frozen segment
+            FrozenObjectHeapManager* foh = SystemDomain::GetFrozenObjectHeapManager();
+            size_t objSize = g_pRuntimeTypeClass->GetBaseSize();
+            refClass = (REFLECTCLASSBASEREF)ObjectToOBJECTREF(foh->TryAllocateObject(g_pRuntimeTypeClass, objSize));
+        }
+
+        if (refClass == NULL)
+        {
+            refClass = (REFLECTCLASSBASEREF)AllocateObject(g_pRuntimeTypeClass);
+        }
 
```

Comment:
We should do the same thing as what we have done for string literals: Skip the handle allocation for frozen objects and use a bit on writeable data to tell whether it is frozen object or not.

Reply:
@jkotas I am not sure I fully understand - writeable data seems to be 8 bytes (on x64) just for the pointer to type object, how can I extend it to include extra bits? Did you mean to use some bit in dwFlags?

Reply:
So is `gc` struct

```
    struct
    {
        OBJECTREF value;
        LOADERALLOCATORREF loaderAllocator;
        PTRARRAYREF handleTable;
        PTRARRAYREF handleTableOld;
    } gc;
```
seems to be packed?

Reply:
Ah, `loaderAllocator` is unused most of the time like in string handles if I read it correctly

Reply:
@jkotas Sorry for pinging again, but I've tried again and it feels like I am a bit lost

so what we save into `MethodTable::m_hExposedClassObject` is a pointer-sized slot in a pinned table that is pointing to `RuntimeType` so I am bit lost on where exactly I can save that bit. What's even more difficult to understand for me is why we [add](https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/loaderallocator.cpp#L844) `+ 1` here and then [subtract `1`](https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/loaderallocator.inl#L98) when we obtain `RuntimeType` from that slot - so basically why it's pointing to the next slot?

Reply:
I would help to make the frozen objects mandatory to reduce the complexity.

If we make the frozen objects mandatory, the logic will become pretty simple: if the low bit is not set, `m_hExposedClassObject` would point directly at the frozen RuntimeTypeHandle (ie we would save the indirection compare to what we have today). If the low bit is set, `m_hExposedClassObject` would point at the slot in collectible handle table like it does today.